### PR TITLE
gh-92031: Deoptimize Static Code at Finalization

### PR DESCRIPTION
**Pull request by sweeneyde**
https://github.com/python/cpython/issues/92031

TODO:
- [x] Find a more deterministic test case
- [x] Decide on EXTENDED_ARG_QUICK/EXTENDED_ARG behavior

### CODE REVIEW

**Comment by sweeneyde** [comment]
One remark: this will convert EXTENDED_ARG_QUICK to EXTENDED_ARG for subsequent interpreters using this code, even though EXTENDED_ARG_QUICK is the one emitted by the compiler.

We should probably add a case for `EXTENDED_ARG --> EXTENDED_ARG_QUICK` in `_PyCode_Quicken`. Though we have a choice: do we

1. Emit (unquickened) EXTENDED_ARG in the compiler, to let `deopt_code()` convert it back to that state. This makes unquickened EXTENDED_ARG opcodes a hair slower, but adds consistency between interpreter invocations.
2. Continue emitting EXTENDED_ARG_QUICK. Then most programs will start as EXTENDED_ARG_QUICK, but there will be an inconsistency between before and after the first `Py_FINALIZE()`
3. Add a special case to `pycode_deopt` that keeps EXTENDED_ARG_QUICK as is.

The central issue is that `_PyOpcode_Deopt[]` has two different contradictory meanings in this case: one is "restore to the original" and one is "what to do during tracing".

File: Lib/test/test_embed.py [hunk 1]
```diff
@@ -343,6 +343,14 @@ def test_finalize_structseq(self):
         out, err = self.run_embedded_interpreter("test_repeated_init_exec", code)
         self.assertEqual(out, 'Tests passed\n' * INIT_LOOPS)
 
+    def test_quickened_static_code_gets_unquickened_at_Py_FINALIZE(self):
```
Comment (neonene) [inline Lib/test/test_embed.py:347]: Could you please specify `@support.skip_if_pgo_task` to make the PGO profiling shorter?

<br>
PS: This issue seems not reproducible for me on Win10 as far as I tried:

1. `test_finalize_structseq -R1:50`
2. `test_quickened_static_code_gets_unquickened_at_Py_FINALIZE`
3. self-contained reproducer

on the debug builds of [b733708](https://github.com/python/cpython/tree/b733708ca), [ea2f5bc](https://github.com/python/cpython/tree/ea2f5bc), [ad5e852](https://github.com/python/cpython/tree/ad5e852) and [special_for_iter2](https://github.com/sweeneyde/cpython/tree/f360d65f344e7b1e1b7c4c85369ff70a36800e7a) branches.

Comment (sweeneyde) [reply Lib/test/test_embed.py:347]: Added skip_if_pgo_task.

Re: reproducibility, it's intermittent for me regardless and I don't know precisely what variables lead it to fail sometimes on the second repetition and sometimes on the 38th (etc.). IIRC I measured failure on about 4% of runs on my Win10 machine.

**Comment by markshannon** [comment]
Feel free to add another table, they are only 256 bytes.
Maybe `_PyOpcode_Original` for restoring back to the original?

**Comment by sweeneyde** [comment]
@markshannon is this okay to merge now?

**Comment by markshannon** [comment]
Yes it is

**Comment by vstinner** [comment]
```
        for i in range(50):
            out, err = run("test_repeated_init_exec", code, timeout=60)
```

test_repeated_init_exec() already repeats the code 4 times. What is the point of repeating the test 50 x 4 =  200 times?

**Comment by vstinner** [comment]
> The failure was still intermittent when I tried it, and just 4 runs only made the test fail about 10-15% of the time when I measured.

The Python test suite is run many times per day on each CI job and we have tons of CI jobs. IMO there is no need to loop so many tests inside the test.

On buildbots, it's very common that bugs which is barely possible to trigger manually when I really want to reproduce them... fail all the time on a specific worker because each worker has different timings and a different operating system ;-)

**Pull request merged**